typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -209,6 +277,32 @@ class Protocol ps where
   --
   data NobodyHasAgency (st :: ps)
 
+  -- | Lemma that if the client has agency for a state, there are no
+  -- cases in which the server has agency for the same state.
+  --
+  proofByContradiction_ClientAndServerHaveAgency
+    :: forall (st :: ps).
+      ClientHasAgency st
+    -> ServerHasAgency st
+    -> Void
+
+  -- | Lemma that if the nobody has agency for a state, there are no
+  -- cases in which the client has agency for the same state.
+  --
+  proofByContradiction_NobodyAndClientHaveAgency
+    :: forall (st :: ps).
+      NobodyHasAgency st
+    -> ClientHasAgency st
+    -> Void
+
+  -- | Lemma that if the nobody has agency for a state, there are no
+  -- cases in which the server has agency for the same state.
+  --
+  proofByContradiction_NobodyAndServerHaveAgency
```

Comment:
I think the names for these 3 should be slightly different. `ClientAndServerHaveAgency` kind of suggests that they both have agency. How about `proofByContradiction_ClientServerExclusive`?

Reply:
Right, I was trying to find a short way of saying "nobody and server have agency leads to a contradition". I'll merge and do some renaming after.